src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs

```diff
@@ -1090,61 +1079,113 @@ private static NamedTypeSymbol GetDelegateType(BoundExpression expr)
+            {
+                return analyzedArguments.Arguments.ToImmutable();
+            }
+
+            int argumentCount = analyzedArguments.Arguments.Count;
+            ArrayBuilder<BoundExpression> newArguments = ArrayBuilder<BoundExpression>.GetInstance(argumentCount);
+            newArguments.AddRange(analyzedArguments.Arguments);
+            for (int i = 0; i < argumentCount; i++)
+            {
+                var argument = newArguments[i];
+                if (argument.Kind == BoundKind.UnboundLambda)
+                {
+                    // bind the argument against each applicable parameter
+                    var unboundArgument = (UnboundLambda)argument;
+                    foreach (var parameterList in parameterListList)
+                    {
+                        var parameterType = GetCorrespondingParameterType(analyzedArguments, i, parameterList);
+                        if (parameterType?.Kind == SymbolKind.NamedType)
                         {
-                            switch (oldArgument.Kind)
-                            {
-                                case BoundKind.UnboundLambda:
-                                    NamedTypeSymbol parameterType = parameters[i].Type as NamedTypeSymbol;
-                                    if ((object)parameterType != null)
-                                    {
-                                        newArguments[i] = ((UnboundLambda)oldArgument).Bind(parameterType);
-                                    }
-                                    break;
-                            }
+                            var discarded = unboundArgument.Bind((NamedTypeSymbol)parameterType);
                         }
-
-                        i++;
                     }
-                    while (i < argumentCount);
 
-                    return newArguments.ToImmutableAndFree();
+                    // replace the unbound lambda with its best inferred bound version
+                    newArguments[i] = unboundArgument.BindForErrorRecovery();
+                }
+            }
+
+            return newArguments.ToImmutableAndFree();
+        }
+
+        /// <summary>
+        /// Compute the type of the corresponding parameter, if any. This is used to improve error recovery,
+        /// for bad invocations, not for semantic analysis of correct invocations, so it is a heuristic.
+        /// If no parameter appears to correspond to the given argument, we return null.
+        /// </summary>
+        /// <param name="analyzedArguments">The analyzed argument list</param>
+        /// <param name="i">The index of the argument</param>
+        /// <param name="parameterList">The parameter list to match against</param>
+        /// <returns>The type of the corresponding parameter.</returns>
+        private static TypeSymbol GetCorrespondingParameterType(AnalyzedArguments analyzedArguments, int i, ImmutableArray<ParameterSymbol> parameterList)
```

Comment:
The code you're looking for is in `BindExtensionMethod`:

```cs
                if (actualArguments == null)
                {
                    // Create a set of arguments for overload resolution of the
                    // extension methods that includes the "this" parameter.
                    actualArguments = AnalyzedArguments.GetInstance();
                    CombineExtensionMethodArguments(left, analyzedArguments, actualArguments);
                }
```

That method already has the comment you're requesting, as you can see.

I'm not concerned about named arguments binding to the first parameter. Any binding we can apply is better than none. I think a user would have to try very hard to get a counterintuitive result in intellisense. This is all heuristic in any case, and until evidence otherwise, I suspect we've arrived at the point of diminishing returns on effort in this area.